src/routes/manage_job/post.rs

```diff
@@ -0,0 +1,294 @@
+//! This file implements the `POST /bulk` endpoint.
+use crate::{
+	errors::ReacherError,
+	routes::check_email::{header::check_header, post::RetryOption},
+};
+use async_recursion::async_recursion;
+use check_if_email_exists::{
+	check_email as ciee_check_email, CheckEmailInput, CheckEmailInputProxy, CheckEmailOutput,
+	Reachable,
+};
+use sqlx::{Pool, Postgres};
+use warp::Filter;
+
+use std::{cmp::min, error::Error};
+
+use crate::routes::check_email::known_errors;
+use serde::{Deserialize, Serialize};
+use sqlxmq::{job, CurrentJob};
+
+const EMAIL_TASK_BATCH_SIZE: usize = 1;
+
+/// Errors that can happen during an email verification.
+#[derive(Debug)]
+enum CheckEmailError {
+	/// We get an `is_reachable` Unknown. We consider this internally as an
+	/// error case, so that we can do retry mechanisms (see select_ok & retry).
+	Unknown((CheckEmailOutput, RetryOption)),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TaskInput {
+	job_id: i32,
+	input: CheckEmailInput,
+}
+
+/// Endpoint request body.
+#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+pub struct CreateBulkRequestBody {
+	input_type: String,
+	input: Vec<String>,
+	proxy: Option<CheckEmailInputProxy>,
+	hello_name: Option<String>,
+	from_email: Option<String>,
+	smtp_port: Option<u16>,
+}
+
+pub struct CreateBulkRequestBodyIterator {
+	body: CreateBulkRequestBody,
+	index: usize,
+	batch_size: usize,
+}
+
+impl IntoIterator for CreateBulkRequestBody {
+	type Item = CheckEmailInput;
+	type IntoIter = CreateBulkRequestBodyIterator;
+
+	fn into_iter(self) -> Self::IntoIter {
+		CreateBulkRequestBodyIterator {
+			body: self,
+			index: 0,
+			batch_size: EMAIL_TASK_BATCH_SIZE,
+		}
+	}
+}
+
+impl Iterator for CreateBulkRequestBodyIterator {
+	type Item = CheckEmailInput;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		if self.index < self.body.input.len() {
+			let bounded_index = min(self.index + self.batch_size, self.body.input.len());
+			let to_emails = self.body.input[self.index..bounded_index].to_vec();
```

Comment:
What is desired here is that each `CheckEmailInput` corresponds to exactly 1 email. So `to_emails` should have exactly 1 item.

The reason is that `check_email()` (which takes CheckEmailInput as input) will output a `Vec<CheckEmailOutput>` with the same length as `to_emails` here, where each item in the array corresponds to the result of the email at the same index.

See L205, we `pop()` on the `Vec<CheckEmailOutput>`, because we assume that there's only 1 item in the output. That assumption only holds if `to_emails`'s length is also 1.

(Under the hood, `check_email()` will check all given emails in parallel, if there are multiple `to_emails`. The whole point of this PR is to switch to a queue model. Hence limiting `check_email()` to only handle 1 email at a time, to have more control on the queue).

Reply:
Ideally, what we want is:
- From CreateBulkRequestBody, create an iterator
- on each iteration, it will read `N` emails from the `CreateBulkRequestBody.input` field
- create `N` distinct `CheckEmailInput`s from these emails
- send these `N` inputs to TaskInput and to sqlxmq
- repeat the iteration

Reply:
I'm wondering if we should tweak `N` ourselves. I though iterators were smart enough to find the optimal `N`.

For example, would the following piece of code work?

```rust
async fn create_bulk_request(
	body: CreateBulkRequestBody,
	conn_pool: Pool<Postgres>,
) -> Result<impl warp::Reply, warp::Rejection> {
	// -- snip --
	for task in body.input.into_iter().map(|email| {
		let mut input = CheckEmailInput::new(vec![email]);
		if let Some(name) = body.hello_name {
			input.set_hello_name(name);
		}

		if let Some(email) = body.from_email {
			input.set_from_email(email);
		}

		if let Some(port) = body.smtp_port {
			input.set_smtp_port(port);
		}

		if let Some(proxy) = body.proxy {
			input.set_proxy(proxy);
		}

		TaskInput {
			input: input,
			job_id: rec.id,
		}
	}) {
		// TODO handle errors gracefully
		email_verification_task
			.builder()
			.set_json(&task)
			.unwrap()
			.spawn(&conn_pool)
			.await
			.unwrap();
	}
}
```

in which case we wouldn't need `CreateBulkRequestBodyIterator`, and we wouldn't handle the `batch_size` ourselves, but let Rust handle it I guess

Reply:
Just for the record, we'll need a separate iterator because this iteration will consume a field in the body of the struct and then still need a reference to struct inside the closure, which the compiler wouldn't allow.